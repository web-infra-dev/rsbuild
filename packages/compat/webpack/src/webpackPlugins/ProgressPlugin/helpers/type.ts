import type { Colors } from '@rsbuild/shared';

export type Props = {
  total: number;
  current: number;
  color: Colors;
  bgColor: Colors;
  char: string;
  width: number;
  buildIcon: string;
  errorIcon: string;
  errorInfo: string;
  message: string;
  done: boolean;
  messageWidth: number;
  spaceWidth: number;
  messageColor: Colors;
  id: string;
  maxIdLen: number;
  hasErrors: boolean;
};

export type BusOption = {
  state: Props[];
};
