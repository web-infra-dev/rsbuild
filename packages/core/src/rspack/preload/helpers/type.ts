import type { HtmlRspackPlugin } from '../../../types';

export type BeforeAssetTagGenerationHtmlPluginData = {
  assets: {
    publicPath: string;
    js: Array<string>;
    css: Array<string>;
    favicon?: string;
    manifest?: string;
  };
  outputName: string;
  plugin: HtmlRspackPlugin;
};

export type As =
  | 'audio'
  | 'document'
  | 'embed'
  | 'fetch'
  | 'font'
  | 'image'
  | 'object'
  | 'script'
  | 'style'
  | 'track'
  | 'worker'
  | 'video';
