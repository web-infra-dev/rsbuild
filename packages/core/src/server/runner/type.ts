import type { RspackOptionsNormalized } from '@rspack/core';

export type RunnerRequirer = (
  currentDirectory: string,
  modulePath: string[] | string,
  context?: {
    file?: BasicRunnerFile;
    esmMode?: EsmMode;
  },
) => Record<string, any> | Promise<Record<string, any>>;

export type BasicRunnerFile = {
  path: string;
  content: string;
  subPath: string;
};

export enum EsmMode {
  Unknown = 0,
  Evaluated = 1,
  Unlinked = 2,
}

export interface BasicModuleScope {
  console: Console;
  [key: string]: any;
}

export interface BasicGlobalContext {
  console: Console;
  setTimeout: typeof setTimeout;
  clearTimeout: typeof clearTimeout;
  [key: string]: any;
}

export type ModuleObject = { exports: unknown };

export type CompilerOptions = RspackOptionsNormalized;

export interface Runner {
  run(file: string): Promise<unknown>;
  getRequire(): RunnerRequirer;
}

export interface RunnerFactory {
  create(
    compilerOptions: CompilerOptions,
    dist: string,
    readFileSync: (fileName: string) => string,
  ): Runner;
}
