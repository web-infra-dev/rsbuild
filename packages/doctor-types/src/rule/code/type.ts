import { RuleErrorCodes } from '.';

/**
 * only defined the code which can be enumerated
 */
export enum RuleMessageCodeEnumerated {
  /**
   * others tools / frameworks errors, such as PIA
   */
  Extend = 'EXTEND',
  /**
   * errors show in the overlay at the client, such as Webpack compile errors in development
   */
  Overlay = 'OVERLAY',
}

export type RuleMessageCode =
  | keyof RuleErrorCodes // enumerated errors
  | `${RuleMessageCodeEnumerated}`
  | RuleMessageCodeEnumerated;

export enum RuleMessageCategory {
  Compile = 'compile',
  Bundle = 'bundle',
  EMO = 'emo',
}

export interface RuleMessage {
  code: RuleMessageCode;
  title: string;
  type: 'text' | 'markdown';
  category: `${RuleMessageCategory}` | RuleMessageCategory;
  description: string;
}
