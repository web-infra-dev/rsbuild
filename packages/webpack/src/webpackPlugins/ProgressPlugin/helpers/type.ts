import type { chalk } from '@rsbuild/shared/chalk';

type Color = (typeof chalk)['ForegroundColor'];

export type Props = {
  total: number;
  current: number;
  color: Color;
  bgColor: Color;
  char: string;
  width: number;
  buildIcon: string;
  message: string;
  done: boolean;
  messageWidth: number;
  spaceWidth: number;
  messageColor: Color;
  id: string;
  maxIdLen: number;
  hasErrors: boolean;
};

export type BusOption = {
  state: Props[];
};
