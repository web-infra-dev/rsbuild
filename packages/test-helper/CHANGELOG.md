# @rsbuild/test-helper

## 0.0.7

## 0.0.6

## 0.0.5

## 0.0.4

### Patch Changes

- 7fca922: feat: setup @rsbuild/test-helper
