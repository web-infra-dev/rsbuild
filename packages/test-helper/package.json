{
  "name": "@rsbuild/test-helper",
  "version": "0.0.27",
  "description": "The vitest helpers of Rsbuild.",
  "homepage": "https://rsbuild.dev",
  "repository": {
    "type": "git",
    "url": "https://github.com/web-infra-dev/rsbuild",
    "directory": "packages/test-helper"
  },
  "license": "MIT",
  "exports": {
    ".": {
      "types": "./dist/index.d.js",
      "default": "./dist/index.js"
    }
  },
  "main": "./dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "modern build",
    "dev": "modern build --watch"
  },
  "dependencies": {
    "@rsbuild/core": "workspace:*",
    "@rsbuild/shared": "workspace:*",
    "@rspack/core": "0.3.14",
    "@types/lodash": "^4.14.200",
    "@types/node": "^16",
    "lodash": "^4.17.21",
    "memfs": "3.4.10",
    "typescript": "^5.2.2",
    "upath": "2.0.1",
    "webpack": "^5.89.0"
  },
  "publishConfig": {
    "access": "public",
    "provenance": true,
    "registry": "https://registry.npmjs.org/",
    "types": "./dist/index.d.ts"
  }
}
