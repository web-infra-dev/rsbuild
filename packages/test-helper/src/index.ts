import { isPathString, normalizeToPosixPath } from './path';
import {
  applyMatcherReplacement,
  createDefaultPathMatchers,
  PathMatcher,
} from './pathSerializer';

export { normalizeToPosixPath } from './path';

export interface SnapshotSerializerOptions {
  cwd?: string;
  workspace?: string;
  replace?: PathMatcher[];
}

export function createSnapshotSerializer(options?: SnapshotSerializerOptions) {
  const {
    cwd = process.cwd(),
    workspace = process.cwd(),
    replace: customMatchers = [],
  } = options || {};

  const pathMatchers: PathMatcher[] = [
    { mark: 'root', match: cwd },
    { mark: 'workspace', match: workspace },
    ...customMatchers,
    ...createDefaultPathMatchers(),
  ];

  pathMatchers
    .filter((matcher) => typeof matcher.match === 'string')
    .forEach(
      (matcher) =>
        (matcher.match = normalizeToPosixPath(matcher.match as string)),
    );

  return {
    pathMatchers,
    // match path-format string
    test: (val: unknown) => typeof val === 'string' && isPathString(val),
    print: (val: unknown) => {
      const normalized = normalizeToPosixPath(val as string);
      const replaced = applyMatcherReplacement(
        pathMatchers,
        normalized,
      ).replace(/"/g, '\\"');
      return `"${replaced}"`;
    },
  };
}

export * from './rsbuild';
