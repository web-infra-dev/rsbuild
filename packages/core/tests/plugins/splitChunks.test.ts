import { createStubRsbuild } from '@rsbuild/test-helper';
import {
  pluginSplitChunks,
  MODULE_PATH_REGEX,
  getPackageNameFromModulePath,
} from '@src/plugins/splitChunks';
import type { RsbuildConfig } from '@rsbuild/shared';

describe('plugin-split-chunks', () => {
  const cases: Array<{ name: string; rsbuildConfig: RsbuildConfig }> = [
    {
      name: 'should set split-by-experience config',
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-experience',
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set split-by-experience config correctly when polyfill is off',
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-experience',
          },
        },
        output: {
          polyfill: 'off',
        },
      },
    },
    {
      name: 'should set split-by-module config',
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-module',
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set single-vendor config',
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'single-vendor',
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set single-size config',
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'split-by-size',
            minSize: 1000,
            maxSize: 5000,
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set all-in-one config',
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'all-in-one',
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should set custom config',
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'custom',
            forceSplitting: [/react/],
            splitChunks: {
              cacheGroups: {},
            },
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should allow forceSplitting to be an object',
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'custom',
            forceSplitting: {
              axios: /axios/,
            },
            splitChunks: {
              cacheGroups: {},
            },
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    },
    {
      name: 'should not split chunks when target is not',
      rsbuildConfig: {
        output: {
          targets: ['node'],
        },
      },
    },
  ];

  it.each(cases)('$name', async (item) => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginSplitChunks()],
      rsbuildConfig: item.rsbuildConfig || {
        performance: {
          chunkSplit: {
            strategy: 'split-by-experience',
          },
        },
        output: {
          polyfill: 'entry',
        },
      },
    });

    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });
});

describe('getPackageNameFromModulePath', () => {
  it('should parse correct path fragment in npm/yarn', async () => {
    let modulePath = '/path/to/node_modules/@scope/package-name/index.js';
    let [, , scope, name] = modulePath.match(MODULE_PATH_REGEX)!;
    expect(scope).toBe('@scope');
    expect(name).toBe('package-name');

    modulePath = '/path/to/node_modules/package-name/index.js';
    [, , scope, name] = modulePath.match(MODULE_PATH_REGEX)!;
    expect(scope).toBe(undefined);
    expect(name).toBe('package-name');
  });

  it('should parse correct path fragment in pnpm', async () => {
    let modulePath = '/path/to/node_modules/.pnpm/@scope/package-name/index.js';
    let [, , scope, name] = modulePath.match(MODULE_PATH_REGEX)!;
    expect(scope).toBe('@scope');
    expect(name).toBe('package-name');

    modulePath = '/path/to/node_modules/.pnpm/package-name/index.js';
    [, , scope, name] = modulePath.match(MODULE_PATH_REGEX)!;
    expect(scope).toBe(undefined);
    expect(name).toBe('package-name');
  });

  it('should return correct package name in npm/yarn', () => {
    let modulePath = '/path/to/node_modules/@scope/package-name/index.js';
    expect(getPackageNameFromModulePath(modulePath)).toBe(
      'npm.scope.package-name',
    );

    modulePath = '/path/to/node_modules/package-name/index.js';
    expect(getPackageNameFromModulePath(modulePath)).toBe('npm.package-name');
  });

  it('should return correct package name in pnpm', () => {
    let modulePath = '/path/to/node_modules/.pnpm/@scope/package-name/index.js';
    expect(getPackageNameFromModulePath(modulePath)).toBe(
      'npm.scope.package-name',
    );

    modulePath = '/path/to/node_modules/.pnpm/package-name/index.js';
    expect(getPackageNameFromModulePath(modulePath)).toBe('npm.package-name');
  });
});
