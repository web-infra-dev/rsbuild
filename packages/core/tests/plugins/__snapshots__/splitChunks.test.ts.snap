// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-split-chunks > 'should allow forceSplitting to be an object' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "bundler-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "axios": {
          "chunks": "all",
          "enforce": true,
          "name": "axios",
          "test": /axios/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;

exports[`plugin-split-chunks > 'should not split chunks when target is not' 1`] = `
{
  "optimization": {
    "splitChunks": false,
  },
}
`;

exports[`plugin-split-chunks > 'should set all-in-one config' 1`] = `
{
  "optimization": {
    "splitChunks": false,
  },
}
`;

exports[`plugin-split-chunks > 'should set custom config' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "bundler-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "force-split-0": {
          "chunks": "all",
          "enforce": true,
          "name": "force-split-0",
          "test": /react/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;

exports[`plugin-split-chunks > 'should set single-size config' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "bundler-runtime",
    },
    "splitChunks": {
      "cacheGroups": {},
      "chunks": "all",
      "enforceSizeThreshold": 50000,
      "maxSize": 5000,
      "minSize": 1000,
    },
  },
}
`;

exports[`plugin-split-chunks > 'should set single-vendor config' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "bundler-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "singleVendor": {
          "chunks": "all",
          "enforce": true,
          "name": "vendor",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;

exports[`plugin-split-chunks > 'should set split-by-experience config correctly when polyfill is off' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "bundler-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "lib-axios": {
          "name": "lib-axios",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(axios\\|axios-\\.\\+\\)\\[\\\\\\\\/\\]/,
        },
        "lib-lodash": {
          "name": "lib-lodash",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(lodash\\|lodash-es\\)\\[\\\\\\\\/\\]/,
        },
        "lib-react": {
          "name": "lib-react",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react\\|react-dom\\|scheduler\\)\\[\\\\\\\\/\\]/,
        },
        "lib-router": {
          "name": "lib-router",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react-router\\|react-router-dom\\|@remix-run\\\\/router\\|history\\)\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;

exports[`plugin-split-chunks > 'should set split-by-experience config' 1`] = `
{
  "optimization": {
    "runtimeChunk": {
      "name": "bundler-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "lib-axios": {
          "name": "lib-axios",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(axios\\|axios-\\.\\+\\)\\[\\\\\\\\/\\]/,
        },
        "lib-lodash": {
          "name": "lib-lodash",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(lodash\\|lodash-es\\)\\[\\\\\\\\/\\]/,
        },
        "lib-polyfill": {
          "name": "lib-polyfill",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(tslib\\|core-js\\|@babel\\\\/runtime\\|@swc\\\\/helpers\\)\\[\\\\\\\\/\\]/,
        },
        "lib-react": {
          "name": "lib-react",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react\\|react-dom\\|scheduler\\)\\[\\\\\\\\/\\]/,
        },
        "lib-router": {
          "name": "lib-router",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react-router\\|react-router-dom\\|@remix-run\\\\/router\\|history\\)\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
}
`;
